### ISSUE TITLE
ASP.NET core 3.1: does ConfigureAppConfiguration interacts with launchSettings.json ?

### ISSUE THREAD

**Issue by EnricoMassone** [dotnet/aspnetcore#30044]
I'm experiencing a strange behavior when launching an ASP.NET core 3.1 web application from Visual Studio 2019 using kestrel (I mean the launching profile which does **not** use IIS express). 

I have created a minimal application to reproduce the issue. 

OS: **Windows 10 (build 19041.746)**
Visual Studio version: **Visual Studio 2019 version 16.8.4**

This is the csproj file:

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>


</Project>
```

This is the launchSettings.json file:

```
{
  "$schema": "http://json.schemastore.org/launchsettings.json",
  "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:52222",
      "sslPort": 0
    }
  },
  "profiles": {
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "launchUrl": "weatherforecast",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    },
    "TestWebApplication": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "weatherforecast",
      "applicationUrl": "http://localhost:5002",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```

This is the Program.cs file:

```
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Hosting;

namespace TestWebApplication
{
  public class Program
  {
    public static void Main(string[] args)
    {
      CreateHostBuilder(args).Build().Run();
    }

    public static IHostBuilder CreateHostBuilder(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureWebHostDefaults(webBuilder =>
            {
              webBuilder.UseStartup<Startup>();
            });
  }
}
```
This is the Startup.cs file:

```
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace TestWebApplication
{
  public class Startup
  {
    public Startup(IConfiguration configuration)
    {
      Configuration = configuration;
    }

    public IConfiguration Configuration { get; }

    // This method gets called by the runtime. Use this method to add services to the container.
    public void ConfigureServices(IServiceCollection services)
    {
      services.AddControllers();
    }

    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
      if (env.IsDevelopment())
      {
        app.UseDeveloperExceptionPage();
      }

      app.UseRouting();

      app.UseAuthorization();

      app.UseEndpoints(endpoints =>
      {
        endpoints.MapControllers();
      });
    }
  }
}
```
When I launch this web application by using the **TestWebApplication** profile from Visual Studio **everything works fine**: kestrel is launched at **http://localhost:5002** and the web browser is launched at **http://localhost:5002/weatherforecast**.
**This is exactly what I would expect from the provided launchSettings.json file.**

Consider the following change to the Program.cs, made to customize the application configuration sources:

```
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;

namespace TestWebApplication
{
  public class Program
  {
    public static void Main(string[] args)
    {
      CreateHostBuilder(args).Build().Run();
    }

    public static IHostBuilder CreateHostBuilder(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureAppConfiguration(builder =>
            {
              builder.Sources.Clear();
              builder.AddEnvironmentVariables();
            })
            .ConfigureWebHostDefaults(webBuilder =>
            {
              webBuilder.UseStartup<Startup>();
            });
  }
}
```

This change breaks everything in terms of interaction with the **launchSettings.json** file. The following happens:

- kester is launched on both **http://localhost:5000** and **https://localhost:5001**
- web browser is **not** launched on **http://localhost:5002/weatherforecast**

Basically it seems that the file **launchSettings.json** is ignored and some defaults are used instead.

I have also observed the following:

- the root cause seems to be the line which clears the configuration sources: `builder.Sources.Clear();`
- the IIS express profile is **not** affected by this issue: it works as expected even with the modified Program.cs file

Is this a bug ? Am I missing anything ?

**Comment by Tratcher** [comment]
This is by design. VS sets the environment variable ASPNETCORE_URLS from applicationUrl. ConfigureWebHostDefaults adds `AddEnvironmentVariables("ASPNETCORE_")` which trims off the prefix, and then Kestrel checks for the "urls" entry and binds to it. The `builder.Sources.Clear();` wipes out the `AddEnvironmentVariables("ASPNETCORE_")` config provider so the data doesn't flow to kestrel as expected.

IIS Express is not affected because the urls are passed to it through a different mechanism.

**Comment by EnricoMassone** [comment]
> This is by design. VS sets the environment variable ASPNETCORE_URLS from applicationUrl. ConfigureWebHostDefaults adds `AddEnvironmentVariables("ASPNETCORE_")` which trims off the prefix, and then Kestrel checks for the "urls" entry and binds to it. The `builder.Sources.Clear();` wipes out the `AddEnvironmentVariables("ASPNETCORE_")` config provider so the data doesn't flow to kestrel as expected.
> 
> IIS Express is not affected because the urls are passed to it through a different mechanism.

@Tratcher is there a way to achieve the following result:

1. fully customize the application configuration by chosing the configuration sources and their order
2. do **not** lose the ability to customize the localhost debugging via `launchSettings.json`

**Comment by Tratcher** [comment]
Add back `AddEnvironmentVariables("ASPNETCORE_")` when building your custom config.

**Comment by EnricoMassone** [comment]
> Add back `AddEnvironmentVariables("ASPNETCORE_")` when building your custom config.

@Tratcher thanks ! It should do the trick. 

Is it needed for development only right ? Based on what you said before it seems this only affects visual studio behavior.

**Comment by Tratcher** [comment]
That depends on your deployment story, the same variable can be used in production depending on your environment.

**Comment by EnricoMassone** [comment]
> That depends on your deployment story, the same variable can be used in production depending on your environment.

@Tratcher thanks for the clarification.

[This](https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs#L38) should be the point where the ASPNETCORE_ environment variables are added to the available configuration.

Actually the environment (development vs production) is determined by one of those variables: ASPNETCORE_ENVIRONMENT.
